### PR TITLE
Alive Robots in robot_factory_position

### PR DESCRIPTION
**Pull request by shourikb**
Changed a few things so that robot_factory_position and a couple other classes use the alive_robots_ array properly. Had to change the type of alive_robots_ from vector to array in a few classes.

In the loop in robot_factory_position, there is a variable j that keeps track of the count and the for each loop gets all of the booleans from alive_robots_ and uses that value in whether to execute the code in the loop body or not.

### CODE REVIEW

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp [hunk 1]
```diff
@@ -69,7 +69,7 @@ class RobotFactoryPosition : public Position {
         current_position_->update_field_dimensions(field_dimensions);
     }
 
-    void update_alive_robots(std::vector<u_int8_t> alive_robots) override {
+    void update_alive_robots(std::array<bool, kNumShells> alive_robots) override {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp:72]:
```suggestion
    void update_alive_robots(std::array<bool, kNumShells> alive_robots) override {
        Position::update_alive_robots(alive_robots);
        current_position->update_alive_robots(alive_robots);
```
You need to call super function

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 2]
```diff
@@ -123,4 +128,8 @@ std::string RobotFactoryPosition::get_current_state() {
     return current_position_->get_current_state();
 }
 
+// void RobotFactoryPosition::update_alive_robots(std::array<bool, kNumShells> alive_robots) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]: delete this commented out code

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 3]
```diff
@@ -26,14 +26,19 @@ std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_stat
     using RobotPos = std::pair<int, double>;  // (robotId, yPosition)
 
     std::vector<RobotPos> robots_copy;
-    for (int i = 0; i < 6; i++) {
+    int j = 0;
+    for (bool alive : alive_robots_) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]:
```suggestion
    for (int i = 0; i < kNumShells; ++i) {
```

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 4]
```diff
@@ -26,14 +26,19 @@ std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_stat
...
         // Ignore goalie
-        if (i == 0) {
+        if (j == goalie_id_) {
             continue;
         }
-        robots_copy.emplace_back(i, world_state.our_robots[i].pose.position().y());
+        if (alive) {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]:
```suggestion
        if (alive_robots_[i]) {
```

Comment (sid-parikh) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]: this is a lot cleaner

**Review by sid-parikh — CHANGES_REQUESTED** [review]
You also need to update the `position.cpp` implementation of alive robots.  It currently searches the list for the robot id, but it should instead check if `alive_robots[robot_id] == true` or something like that, because you changed it from a vector to an array

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp (hunk 1 above)
Comment (sid-parikh) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp:72]: no, that's not correct. you need to call the super function _for robot factory position_, and then call the current_position function _without Position::_

Comment (sid-parikh) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp:72]: I've updated my GH suggestion to show what I mean

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp [hunk 5]
```diff
@@ -69,8 +69,8 @@ class RobotFactoryPosition : public Position {
         current_position_->update_field_dimensions(field_dimensions);
     }
 
-    void update_alive_robots(std::vector<u_int8_t> alive_robots) override {
-        current_position_->update_alive_robots(alive_robots);
+    void update_alive_robots(std::array<bool, kNumShells> alive_robots) override {
+        current_position_->Position::update_alive_robots(alive_robots);
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.hpp]:
```suggestion

```

**Review by sid-parikh — CHANGES_REQUESTED** [review]

File: soccer/src/soccer/strategy/agent/agent_action_client.cpp [hunk 6]
```diff
@@ -116,6 +116,8 @@ void AgentActionClient::goalie_id_callback(int goalie_id) {
 
 void AgentActionClient::alive_robots_callback(const rj_msgs::msg::AliveRobots::SharedPtr& msg) {
     alive_robots_ = msg->alive_robots;
+
+    current_position_->Position::update_alive_robots(alive_robots_);
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/agent_action_client.cpp]:
```suggestion
    current_position_->update_alive_robots(alive_robots_);
```

File: soccer/src/soccer/radio/sim_radio.cpp [hunk 7]
```diff
@@ -94,18 +94,19 @@ SimRadio::SimRadio(bool blue_team)
         ip::udp::endpoint(address_, blue_team_ ? kSimBlueCommandPort : kSimYellowCommandPort);
     sim_control_endpoint_ = ip::udp::endpoint(address_, kSimCommandPort);
 
+    // assume robots 0-6 are always alive in sim
```
Comment (rishiso) [inline soccer/src/soccer/radio/sim_radio.cpp:97]:
```suggestion
    // assume robots 0-5 are always alive in sim
```

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 8]
```diff
@@ -49,29 +50,29 @@ std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_stat
     // Assigning new position
     // Checking whether we have possesion or if the ball is on their half (using 1.99 to avoid
```
Comment (rishiso) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:51]:
```suggestion
    // Checking whether we have possesion or if the ball is on their half
```

Comment (rishiso) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:51]:
```suggestion
    // Checking whether we have possesion or if the ball is on their half
```

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 9]
```diff
@@ -49,29 +50,29 @@ std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_stat
...
     // rounding issues on midline)
```
Comment (rishiso) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:52]:
```suggestion
```

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 10]
```diff
@@ -6,7 +6,6 @@ RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id, "RobotFact
     if (robot_id_ == 0) {
```
Comment (rishiso) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:6]:
```suggestion
    if (robot_id_ == goalie_id_) {
```

Comment (rishiso) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:6]: Goalie is never reassigned inside get_task(), so it is important to do this here

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 11]
```diff
@@ -16,7 +15,7 @@ RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id, "RobotFact
 std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_state,
                                                           FieldDimensions& field_dimensions) {
     // If keeper, make no changes
-    if (robot_id_ == 0) {
+    if (robot_id_ == goalie_id_) {
```
Comment (rishiso) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:18]: I will need to double-check with Sid about this, but if the goalie can change, we need to add functionality here to update the goalie if it is not assigned correctly.

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 12]
```diff
@@ -49,29 +50,29 @@ std::optional<RobotIntent> RobotFactoryPosition::get_task(WorldState& world_stat
     // Assigning new position
     // Checking whether we have possesion or if the ball is on their half (using 1.99 to avoid
     // rounding issues on midline)
-    if (Position::our_possession_ ||
-        world_state.ball.position.y() > field_dimensions.length() / 1.99) {
+    if (our_possession_ ||
+        world_state.ball.position.y() > field_dimensions.center_field_loc().y() - kBallDiameter) {
         // Offensive mode
         // Closest 2 robots on defense, rest on offense
         if (i <= 1) {
```
Comment (jacksherling) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:57]: While we're working on this file, could we change the i variable name to something more descriptive? Perhaps `robot_y_index`?
Variable defined on line 42 but it's collapsed so GitHub won't let me comment there

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp (hunk 10 above)
Comment (jacksherling) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:6]: good catch

**Review by jacksherling — CHANGES_REQUESTED** [review]

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp (hunk 10 above)
Comment (jacksherling) [reply soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp:6]: good catch

**Comment by sid-parikh** [comment]
Merged into kicker-picker see #2200

**Pull request closed without merging**